### PR TITLE
[FLINK-15935][table] Fix watermark can't work when depending both on flink planner and blink planner in project

### PR DESCRIPTION
**Pull request by wuchong**
## What is the purpose of the change

The reason is that we didn't sync the fixing for org.apache.calcite.sql.validate.ParameterScope to flink-planner. When an application project dependent flink-planner and blink-planner at the same time, the classloader may use the ParameterScope from Calcite intead of from planner, which leads to this exception.

The fixing is simple, just copy the org.apache.calcite.sql.validate.ParameterScope class from blink planner to flink planner.

## Brief change log

- copy `org.apache.calcite.sql.validate.ParameterScope` class from blink planner to flink planner.
- add an `StreamWindowSQLExample` which uses DDL and watermark to reproduce the problem, and also this example is useful for users to learn about the usage of DDL and watermark statement.

## Verifying this change

By running the `StreamWindowSQLExample`.

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (yes / **no**)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (yes / **no**)
  - The serializers: (yes / **no** / don't know)
  - The runtime per-record code paths (performance sensitive): (yes / **no** / don't know)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Yarn/Mesos, ZooKeeper: (yes / **no** / don't know)
  - The S3 file system connector: (yes / **no** / don't know)

## Documentation

  - Does this pull request introduce a new feature? (yes / **no**)
  - If yes, how is the feature documented? (**not applicable** / docs / JavaDocs / not documented)

### CODE REVIEW

**Comment by JingsongLi** [comment]
Thanks @wuchong , please try to solve `org.apache.calcite.sql.validate.SqlValidatorImpl` and `org.apache.calcite.sql2rel.RelDecorrelator`.
They also have potential problems.

**Comment by wuchong** [comment]
Thanks @JingsongLi, I have removed `RelDecorrelator ` because CALCITE-3169 and CALCITE-3170 have already been fixed. And copy `SqlValidatorImpl` to blink planner to throw exception for count distinct in MATCH RECOGNIZE.

File: flink-examples/flink-examples-table/src/main/java/org/apache/flink/table/examples/java/StreamWindowSQLExample.java [hunk 1]
```diff
@@ -0,0 +1,94 @@
...
+
+import java.net.URL;
+import java.nio.file.Paths;
+
+/**
+ * Simple example for demonstrating the use of SQL in Java.
+ *
+ * <p>This example shows how to:
+ *  - Register a table via DDL
+ *  - Declare event time attribute in the DDL
```
Comment (twalthr) [inline flink-examples/flink-examples-table/src/main/java/org/apache/flink/table/examples/java/StreamWindowSQLExample.java]: nit: `an event time attribute`

File: flink-examples/flink-examples-table/src/main/java/org/apache/flink/table/examples/java/StreamWindowSQLExample.java [hunk 2]
```diff
@@ -0,0 +1,94 @@
...
+ *  - Run a streaming window aggregate on the registered table
+ *
```
Comment (twalthr) [inline flink-examples/flink-examples-table/src/main/java/org/apache/flink/table/examples/java/StreamWindowSQLExample.java]: nit: remove empty line

File: flink-examples/flink-examples-table/src/main/java/org/apache/flink/table/examples/java/StreamWindowSQLExample.java [hunk 3]
```diff
@@ -0,0 +1,94 @@
...
+ */
+public class StreamWindowSQLExample {
+
+	// *************************************************************************
```
Comment (twalthr) [inline flink-examples/flink-examples-table/src/main/java/org/apache/flink/table/examples/java/StreamWindowSQLExample.java]: nit: I think we can remove this section if we don't have utility classes.

File: flink-examples/flink-examples-table/src/main/java/org/apache/flink/table/examples/java/StreamWindowSQLExample.java [hunk 4]
```diff
@@ -0,0 +1,94 @@
...
+		// use blink planner in streaming mode,
+		// because watermark statement is only available in blink planner.
+		EnvironmentSettings settings = EnvironmentSettings.newInstance()
+			.useBlinkPlanner()
+			.inStreamingMode()
+			.build();
+		StreamTableEnvironment tEnv = StreamTableEnvironment.create(env, settings);
+
+		// get the absolute path of the source content file
+		URL res = StreamWindowSQLExample.class.getClassLoader().getResource("orders.csv");
```
Comment (twalthr) [inline flink-examples/flink-examples-table/src/main/java/org/apache/flink/table/examples/java/StreamWindowSQLExample.java]: Is this guaranteed to also work on windows?

File: flink-examples/flink-examples-table/src/main/java/org/apache/flink/table/examples/java/StreamWindowSQLExample.java [hunk 5]
```diff
@@ -0,0 +1,94 @@
...
+			"  'connector.type' = 'filesystem',\n" +
+			"  'connector.path' = '" + path + "',\n" +
+			"  'format.type' = 'csv'\n" +
+			")";
+		tEnv.sqlUpdate(ddl);
+
+		// run a SQL query on the table and retrieve the result as a new Table
+		String query = "SELECT \n" +
+			"  CAST(TUMBLE_START(ts, INTERVAL '5' SECOND) AS STRING) window_start, \n" +
+			"  COUNT(*) order_num, \n" +
```
Comment (twalthr) [inline flink-examples/flink-examples-table/src/main/java/org/apache/flink/table/examples/java/StreamWindowSQLExample.java]: nit: sometimes there is a space before the newline sometimes not

File: flink-examples/flink-examples-table/src/main/java/org/apache/flink/table/examples/java/StreamWindowSQLExample.java (hunk 4 above)
Comment (wuchong) [reply flink-examples/flink-examples-table/src/main/java/org/apache/flink/table/examples/java/StreamWindowSQLExample.java]: Using csv file in resources will break the RAT, so I turned to create a temporary file and write the content in it.

File: flink-examples/flink-examples-table/src/main/java/org/apache/flink/table/examples/java/StreamWindowSQLExample.java [hunk 6]
```diff
@@ -0,0 +1,98 @@
...
+
+		// set up execution environment
+		StreamExecutionEnvironment env = StreamExecutionEnvironment.getExecutionEnvironment();
+		// use blink planner in streaming mode,
+		// because watermark statement is only available in blink planner.
+		EnvironmentSettings settings = EnvironmentSettings.newInstance()
+			.useBlinkPlanner()
+			.inStreamingMode()
+			.build();
+		StreamTableEnvironment tEnv = StreamTableEnvironment.create(env, settings);
```
Comment (JingsongLi) [inline flink-examples/flink-examples-table/src/main/java/org/apache/flink/table/examples/java/StreamWindowSQLExample.java]: Can we use `TableEnvironment` and `TableUtils.collectToList`?

Comment (wuchong) [reply flink-examples/flink-examples-table/src/main/java/org/apache/flink/table/examples/java/StreamWindowSQLExample.java]: This is an example, when users submit the job, the query result should display in console or `.out` log files. This is not intended to verify the result.

**Comment by wuchong** [comment]
Thanks @JingsongLi and @twalthr  for the reviewing. 

Merging...

**Pull request closed without merging**